### PR TITLE
feat(collections): Lazy storage value

### PR DESCRIPTION
**Pull request by austinabell**
As mentioned in #406 this implements the lazy storage type, which is similar to `LazyOption` except that the assertion that the value exists in storage is maintained by the API to allow easier usage for developers.

This PR re-uses the cache entry type from #402 (with small adjustments) to solidify that API and ensure it can be used for other types. Also for simplicity, this brings in the `once_cell` crate for lazily filling the cache cell under the feature flag, but only specific logic needed can be ported over to minimize compiled size if necessary.

### CODE REVIEW

**Comment by austinabell** [comment]
per our discussions, I will be moving this type to the new module. The name is still not specified, so either `storage` or `store`, does anyone have opinions?

File: near-sdk/Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,9 @@ near-primitives-core = "=0.4.0"
 # Export dependencies for contracts
 wee_alloc = { version = "0.4.5", default-features = false, features = [] }
 
+# Used for caching, might be worth porting only functionality needed.
+once_cell = { version = "1.7.2", optional = true }
```
Comment (matklad) [inline near-sdk/Cargo.toml:33]:
```suggestion
once_cell = { version = "1.7.2", optional = true, default-features = false }
```

I think need only no_std (single threaded ) version.

**Comment by matklad** [comment]
`store` feels better to me -- shorter, and it has a "a specific thing" connotation, which is good -- `store` refers to a specific global store.

**Review by evgenykuzyakov — APPROVED** [review]

File: near-sdk/src/store/lazy/impls.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use super::Lazy;
```
Comment (mikedotexe) [inline near-sdk/src/store/lazy/impls.rs:3]: me this weekend.

File: near-sdk/src/store/lazy/mod.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    let serialized = value.try_to_vec().unwrap_or_else(|_| env::panic(ERR_VALUE_SERIALIZATION));
+    env::storage_write(&key, &serialized);
+}
+
+/// An persistent lazily loaded value, that stores a value in the storage.
+///
+/// This will only write to the underlying store if the value has changed, and will only read the
+/// existing value from storage once.
+///
+/// # Examples
```
Comment (mikedotexe) [inline near-sdk/src/store/lazy/mod.rs:52]: Again appreciate these doc examples

**Review by mikedotexe — APPROVED** [review]

**Pull request merged**